File: bindings/python/flowd/trace.py
"""Read and diff Flow execution traces.

A trace is the durable, content-addressed record of one flow run — every
node's inputs, outputs, model identities, token/cost metrics, and timing
(see flow-trace(5)). This module turns a trace directory into an inspectable
object graph and diffs two traces node-by-node, which is what makes the
record/replay/diff loop usable from Python:

    base    = Runtime(Impl()).run(x)            # records traces/<flow>/<exec>/
    replay  = Runtime.open(ir).replay(base_dir, new_dir, model="gpt-5")
    delta   = diff_traces(base_dir, new_dir)    # what changed, and where
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Optional


def _resolve(value: Any, values_dir: Path) -> Any:
    """Resolve a trace value reference: {"inline": v} | {"hash": "sha256:.."}."""
    # Safe only because the trace writer wraps every value in exactly one of
    # {inline} / {hash}: a one-key dict here is a reference, never a user value.
    # If that wrapping ever loosens, a real {"hash": ..} payload would be
    # misread as a content address.
    if isinstance(value, dict) and len(value) == 1:
        if "inline" in value:
            return value["inline"]
        if "hash" in value:
            digest = str(value["hash"]).split(":", 1)[-1]
            blob = values_dir / f"{digest}.json"
            return json.loads(blob.read_text())
    return value


@dataclass
class Invocation:
    """One row of input a node processed (a direct call has exactly one)."""

    inputs: Optional[dict]
    output: Any


@dataclass
class Node:
    """One node's execution record."""

    node_id: str
    node_kind: str          # input | output | tool_call | model_call | subflow_call | suspension
    effect_level: str       # pure | deterministic | model | mutation
    invocations: list[Invocation]
    tool: Optional[str] = None
    provider: Optional[str] = None
    model: Optional[str] = None
    model_version: Optional[str] = None
    tokens_in: int = 0
    tokens_out: int = 0
    cost_cents: float = 0.0
    elapsed_ms: int = 0
    errors: Optional[dict] = None
    raw: dict = field(default_factory=dict, repr=False)

    @property
    def output(self) -> Any:
        """The node's produced value (last invocation's output)."""
        return self.invocations[-1].output if self.invocations else None

    @property
    def failed(self) -> bool:
        return self.errors is not None


class Trace:
    """A parsed trace directory (manifest + nodes + content-addressed values)."""

    def __init__(self, directory: str | Path):
        self.dir = Path(directory)
        manifest_path = self.dir / "manifest.json"
        if not manifest_path.is_file():
            raise FileNotFoundError(f"no manifest.json under {self.dir} (R301)")
        self.manifest: dict = json.loads(manifest_path.read_text())

        values_dir = self.dir / "values"
        self.nodes: list[Node] = []
        nodes_dir = self.dir / "nodes"
        if nodes_dir.is_dir():
            for path in sorted(nodes_dir.glob("*.json"), key=lambda p: _id_sort(p.stem)):
                self.nodes.append(_parse_node(json.loads(path.read_text()), values_dir))
        self._by_id = {n.node_id: n for n in self.nodes}

    # -- manifest fields ---------------------------------------------------
    @property
    def execution_id(self) -> str:
        return self.manifest.get("execution_id", "")

    @property
    def flow(self) -> str:
        return self.manifest.get("flow", "")

    @property
    def status(self) -> str:
        return self.manifest.get("status", "")

    @property
    def ir_hash(self) -> str:
        return self.manifest.get("ir_hash", "")

    @property
    def input_hash(self) -> str:
        return self.manifest.get("input_hash", "")

    @property
    def output_hash(self) -> str:
        return self.manifest.get("output_hash", "")

    @property
    def elapsed_ms(self) -> int:
        return self.manifest.get("elapsed_ms", 0)

    @property
    def budget(self) -> dict:
        """Run totals: {'tokens_in', 'tokens_out', 'cost_cents'}."""
        return self.manifest.get("budget_summary", {})

    @property
    def model_calls(self) -> list[Node]:
        return [n for n in self.nodes if n.node_kind == "model_call"]

    @property
    def input(self) -> Any:
        n = next((n for n in self.nodes if n.node_kind == "input"), None)
        return n.output if n else None

    @property
    def output(self) -> Any:
        n = next((n for n in self.nodes if n.node_kind == "output"), None)
        return n.output if n else None

    def node(self, node_id: str) -> Optional[Node]:
        return self._by_id.get(node_id)

    def __repr__(self) -> str:
        return (
            f"Trace({self.flow}/{self.execution_id} status={self.status} "
            f"nodes={len(self.nodes)} budget={self.budget})"
        )


def _parse_node(raw: dict, values_dir: Path) -> Node:
    invocations = []
    for inv in raw.get("invocations", []):
        inputs = inv.get("inputs")
        if isinstance(inputs, dict):
            inputs = {k: _resolve(v, values_dir) for k, v in inputs.items()}
        output = _resolve(inv["output"], values_dir) if "output" in inv else None
        invocations.append(Invocation(inputs=inputs, output=output))
    return Node(
        node_id=raw["node_id"],
        node_kind=raw["node_kind"],
        effect_level=raw.get("effect_level", ""),
        invocations=invocations,
        tool=raw.get("tool"),
        provider=raw.get("provider"),
        model=raw.get("model"),
        model_version=raw.get("model_version"),
        tokens_in=raw.get("tokens_in", 0),
        tokens_out=raw.get("tokens_out", 0),
        cost_cents=raw.get("cost_cents", 0.0),
        elapsed_ms=raw.get("elapsed_ms", 0),
        errors=raw.get("errors"),
        raw=raw,
    )


def open_trace(directory: str | Path) -> Trace:
    """Open a trace directory (the traces/<flow>/<execution_id>/ produced by a run)."""
    return Trace(directory)


# -- diff -------------------------------------------------------------------

@dataclass
class NodeChange:
    node_id: str
    node_kind: str
    output_before: Any = None
    output_after: Any = None
    model_before: Optional[str] = None
    model_after: Optional[str] = None

    @property
    def output_changed(self) -> bool:
        return self.output_before != self.output_after

    @property
    def model_changed(self) -> bool:
        return self.model_before != self.model_after


@dataclass
class TraceDiff:
    """Node-by-node difference between two traces (same flow, e.g. a run and
    its model-versioned replay)."""

    changes: list[NodeChange] = field(default_factory=list)
    only_in_a: list[str] = field(default_factory=list)
    only_in_b: list[str] = field(default_factory=list)
    output_hash_before: str = ""
    output_hash_after: str = ""

    @property
    def output_changed(self) -> bool:
        return self.output_hash_before != self.output_hash_after

    def __bool__(self) -> bool:
        return bool(self.changes or self.only_in_a or self.only_in_b)

    def summary(self) -> str:
        if not self:
            return "traces identical"
        lines = [f"{len(self.changes)} node(s) changed; "
                 f"output {'changed' if self.output_changed else 'unchanged'}"]
        for c in self.changes:
            bits = []
            if c.output_changed:
                bits.append(f"output {c.output_before!r} -> {c.output_after!r}")
            if c.model_changed:
                bits.append(f"model {c.model_before} -> {c.model_after}")
            lines.append(f"  {c.node_id} ({c.node_kind}): {'; '.join(bits)}")
        if self.only_in_a:
            lines.append(f"  only in A: {', '.join(self.only_in_a)}")
        if self.only_in_b:
            lines.append(f"  only in B: {', '.join(self.only_in_b)}")
        return "\n".join(lines)

    def __repr__(self) -> str:
        return f"TraceDiff(changes={len(self.changes)}, output_changed={self.output_changed})"


def diff_traces(a: str | Path | Trace, b: str | Path | Trace) -> TraceDiff:
    """Diff two traces node-by-node, by node id."""
    ta = a if isinstance(a, Trace) else open_trace(a)
    tb = b if isinstance(b, Trace) else open_trace(b)

    ids_a = {n.node_id for n in ta.nodes}
    ids_b = {n.node_id for n in tb.nodes}
    diff = TraceDiff(
        only_in_a=sorted(ids_a - ids_b, key=_id_sort),
        only_in_b=sorted(ids_b - ids_a, key=_id_sort),
        output_hash_before=ta.output_hash,
        output_hash_after=tb.output_hash,
    )
    for node_id in sorted(ids_a & ids_b, key=_id_sort):
        na, nb = ta.node(node_id), tb.node(node_id)
        change = NodeChange(
            node_id=node_id,
            node_kind=na.node_kind,
            output_before=na.output,
            output_after=nb.output,
            model_before=na.model,
            model_after=nb.model,
        )
        if change.output_changed or change.model_changed:
            diff.changes.append(change)
    return diff


def _id_sort(node_id: str):
    return (0, int(node_id[1:])) if node_id[1:].isdigit() else (1, node_id)

File: bindings/typescript/src/trace.ts
/**
 * Read and diff Flow execution traces.
 *
 * A trace is the durable, content-addressed record of one flow run — every
 * node's inputs, outputs, model identities, token/cost metrics, and timing
 * (flow-trace(5)). This turns a trace directory into an inspectable object
 * graph and diffs two traces node-by-node, making the record/replay/diff loop
 * usable from TypeScript.
 */

import * as fs from "node:fs";
import * as path from "node:path";

function resolveValue(value: unknown, valuesDir: string): unknown {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    const o = value as Record<string, unknown>;
    const keys = Object.keys(o);
    if (keys.length === 1) {
      if ("inline" in o) return o.inline;
      if ("hash" in o) {
        const digest = String(o.hash).split(":").pop();
        return JSON.parse(fs.readFileSync(path.join(valuesDir, `${digest}.json`), "utf8"));
      }
    }
  }
  return value;
}

export interface Invocation {
  inputs: Record<string, unknown> | null;
  output: unknown;
}

export interface Node {
  nodeId: string;
  nodeKind: string; // input | output | tool_call | model_call | subflow_call | suspension
  effectLevel: string; // pure | deterministic | model | mutation
  invocations: Invocation[];
  tool?: string;
  provider?: string;
  model?: string;
  modelVersion?: string;
  tokensIn: number;
  tokensOut: number;
  costCents: number;
  elapsedMs: number;
  errors: unknown | null;
  /** The node's produced value (last invocation's output). */
  output: unknown;
  failed: boolean;
}

interface Manifest {
  execution_id?: string;
  flow?: string;
  status?: string;
  ir_hash?: string;
  input_hash?: string;
  output_hash?: string;
  elapsed_ms?: number;
  budget_summary?: { tokens_in: number; tokens_out: number; cost_cents: number };
}

function numKey(id: string): number {
  return id[0] === "n" && /^\d+$/.test(id.slice(1)) ? Number(id.slice(1)) : Infinity;
}

function parseNode(raw: any, valuesDir: string): Node {
  const invocations: Invocation[] = (raw.invocations ?? []).map((inv: any) => {
    let inputs: Record<string, unknown> | null = inv.inputs ?? null;
    if (inputs && typeof inputs === "object") {
      inputs = Object.fromEntries(
        Object.entries(inputs).map(([k, v]) => [k, resolveValue(v, valuesDir)]),
      );
    }
    return {
      inputs,
      output: "output" in inv ? resolveValue(inv.output, valuesDir) : null,
    };
  });
  return {
    nodeId: raw.node_id,
    nodeKind: raw.node_kind,
    effectLevel: raw.effect_level ?? "",
    invocations,
    tool: raw.tool,
    provider: raw.provider,
    model: raw.model,
    modelVersion: raw.model_version,
    tokensIn: raw.tokens_in ?? 0,
    tokensOut: raw.tokens_out ?? 0,
    costCents: raw.cost_cents ?? 0,
    elapsedMs: raw.elapsed_ms ?? 0,
    errors: raw.errors ?? null,
    output: invocations.length ? invocations[invocations.length - 1].output : null,
    failed: (raw.errors ?? null) !== null,
  };
}

export class Trace {
  readonly dir: string;
  readonly manifest: Manifest;
  readonly nodes: Node[];
  private readonly byId: Map<string, Node>;

  constructor(directory: string) {
    this.dir = directory;
    const manifestPath = path.join(directory, "manifest.json");
    if (!fs.existsSync(manifestPath)) {
      throw new Error(`no manifest.json under ${directory} (R301)`);
    }
    this.manifest = JSON.parse(fs.readFileSync(manifestPath, "utf8"));

    const valuesDir = path.join(directory, "values");
    const nodesDir = path.join(directory, "nodes");
    this.nodes = fs.existsSync(nodesDir)
      ? fs
          .readdirSync(nodesDir)
          .filter((f) => f.endsWith(".json"))
          .map((f) => parseNode(JSON.parse(fs.readFileSync(path.join(nodesDir, f), "utf8")), valuesDir))
          .sort((a, b) => numKey(a.nodeId) - numKey(b.nodeId))
      : [];
    this.byId = new Map(this.nodes.map((n) => [n.nodeId, n]));
  }

  get executionId(): string { return this.manifest.execution_id ?? ""; }
  get flow(): string { return this.manifest.flow ?? ""; }
  get status(): string { return this.manifest.status ?? ""; }
  get irHash(): string { return this.manifest.ir_hash ?? ""; }
  get inputHash(): string { return this.manifest.input_hash ?? ""; }
  get outputHash(): string { return this.manifest.output_hash ?? ""; }
  get elapsedMs(): number { return this.manifest.elapsed_ms ?? 0; }
  /** Run totals: { tokens_in, tokens_out, cost_cents }. */
  get budget() { return this.manifest.budget_summary ?? { tokens_in: 0, tokens_out: 0, cost_cents: 0 }; }
  get modelCalls(): Node[] { return this.nodes.filter((n) => n.nodeKind === "model_call"); }
  get input(): unknown { return this.nodes.find((n) => n.nodeKind === "input")?.output ?? null; }
  get output(): unknown { return this.nodes.find((n) => n.nodeKind === "output")?.output ?? null; }

  node(nodeId: string): Node | undefined { return this.byId.get(nodeId); }
  diff(other: Trace): TraceDiff { return diffTraces(this, other); }
}

export function openTrace(directory: string): Trace {
  return new Trace(directory);
}

// ---- diff ------------------------------------------------------------------

export interface NodeChange {
  nodeId: string;
  nodeKind: string;
  outputBefore: unknown;
  outputAfter: unknown;
  modelBefore?: string;
  modelAfter?: string;
  outputChanged: boolean;
  modelChanged: boolean;
}

export class TraceDiff {
  constructor(
    readonly changes: NodeChange[],
    readonly onlyInA: string[],
    readonly onlyInB: string[],
    readonly outputHashBefore: string,
    readonly outputHashAfter: string,
  ) {}

  get outputChanged(): boolean {
    return this.outputHashBefore !== this.outputHashAfter;
  }

  get hasChanges(): boolean {
    return this.changes.length > 0 || this.onlyInA.length > 0 || this.onlyInB.length > 0;
  }

  summary(): string {
    if (!this.hasChanges) return "traces identical";
    const lines = [
      `${this.changes.length} node(s) changed; output ${this.outputChanged ? "changed" : "unchanged"}`,
    ];
    for (const c of this.changes) {
      const bits: string[] = [];
      if (c.outputChanged) bits.push(`output ${JSON.stringify(c.outputBefore)} -> ${JSON.stringify(c.outputAfter)}`);
      if (c.modelChanged) bits.push(`model ${c.modelBefore} -> ${c.modelAfter}`);
      lines.push(`  ${c.nodeId} (${c.nodeKind}): ${bits.join("; ")}`);
    }
    if (this.onlyInA.length) lines.push(`  only in A: ${this.onlyInA.join(", ")}`);
    if (this.onlyInB.length) lines.push(`  only in B: ${this.onlyInB.join(", ")}`);
    return lines.join("\n");
  }
}

export function diffTraces(a: string | Trace, b: string | Trace): TraceDiff {
  const ta = a instanceof Trace ? a : openTrace(a);
  const tb = b instanceof Trace ? b : openTrace(b);
  const idsA = new Set(ta.nodes.map((n) => n.nodeId));
  const idsB = new Set(tb.nodes.map((n) => n.nodeId));

  const changes: NodeChange[] = [];
  for (const id of [...idsA].filter((x) => idsB.has(x)).sort((x, y) => numKey(x) - numKey(y))) {
    const na = ta.node(id)!;
    const nb = tb.node(id)!;
    const outputChanged = JSON.stringify(na.output) !== JSON.stringify(nb.output);
    const modelChanged = na.model !== nb.model;
    if (outputChanged || modelChanged) {
      changes.push({
        nodeId: id,
        nodeKind: na.nodeKind,
        outputBefore: na.output,
        outputAfter: nb.output,
        modelBefore: na.model,
        modelAfter: nb.model,
        outputChanged,
        modelChanged,
      });
    }
  }
  return new TraceDiff(
    changes,
    [...idsA].filter((x) => !idsB.has(x)).sort((x, y) => numKey(x) - numKey(y)),
    [...idsB].filter((x) => !idsA.has(x)).sort((x, y) => numKey(x) - numKey(y)),
    ta.outputHash,
    tb.outputHash,
  );
}
